pkg/resource/managed_reconciler.go

```diff
@@ -222,15 +222,18 @@ type ManagedReconciler struct {
 
 type mrManaged struct {
 	ManagedConnectionPublisher
-	ManagedEstablisher
 	ManagedFinalizer
+	ManagedInitializer
 }
 
 func defaultMRManaged(m manager.Manager) mrManaged {
 	return mrManaged{
 		ManagedConnectionPublisher: NewAPISecretPublisher(m.GetClient(), m.GetScheme()),
-		ManagedEstablisher:         NewAPIManagedFinalizerAdder(m.GetClient()),
 		ManagedFinalizer:           NewAPIManagedFinalizerRemover(m.GetClient()),
+		ManagedInitializer: InitializerChain{
```

Comment:
The developers who work on the resources that acquires their name after the creation will have to provide their own `Initializer`s. With this pattern, instead of supplying an empty chain, they should supply `NewAPIManagedFinalizerAdder(m.GetClient())`. Not sure how we'd enforce that.

In fact, I started to think why allow people not to use `NewAPIManagedFinalizerAdder(m.GetClient())`? It could be a part of managed reconciler and all resource would be processed by it. I couldn't come up with a case where we don't want to set the finalizer of the resource, given that external policies(or something possibly other than finalizers) will be how we handle the management of the resource #27

Reply:
> Not sure how we'd enforce that.

We could just inject it in the `InitializerChain` within the `WithManagedInitializer()` func. However, that would effectively render it impossible not to use `NewAPIManagedFinalizerAdder(m.GetClient())`.

Reply:
Then if they did use `NewAPIManagedFinalizerAdder(m.GetClient())` in `WithManagedInitializers` call, there would be duplicate of it.

But if we inject it and don't export `APIManagedFinalizerAdder` and `APIManagedFinalizerRemover` and friends at all, they'd not be able to add it. I think I'll go with that.

Reply:
Yeah if we are going use it no matter what then exposing it to the user is a bad UX.

Reply:
Well, `WithManagedInitializers` doesn't have a way to access the client that's needed to call `NewAPIManagedFinalizerAdder`. I'd like to get more feedback on whether it's worth it to enforce `NewAPIManagedFinalizerAdder` even though developer forgets it for the resources that won't use `NewManagedNameAsExternalName`(not too few) before making not-so-small/hacky changes to make it work. @negz

Reply:
Hrm this is a tough one. In practice pretty much every managed resource is going to want to add a finalizer.

I'm inclined to keep this as you currently have it implemented. If your controller falls into the most common case of using its managed resource name you can stick with the default. If you need to override the default behaviour then you need to feed in the FinalizerAdder as well as your external name strategy.